### PR TITLE
HDDS-1517. AllocateBlock call fails with ContainerNotFoundException.

### PR DESCRIPTION
**Pull request by bshashikant**

### CODE REVIEW

File: hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/TestSCMContainerManager.java [hunk 1]
```diff
@@ -43,17 +46,15 @@
...
 import java.io.IOException;
-import java.util.Iterator;
-import java.util.Optional;
-import java.util.Random;
-import java.util.Set;
-import java.util.TreeSet;
-import java.util.UUID;
-import java.util.concurrent.TimeUnit;
+import java.util.*;
+import java.util.concurrent.*;
```
Comment (jiwq) [inline hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/TestSCMContainerManager.java]: Should better don't use `*` in import statement.

File: hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/TestSCMContainerManager.java [hunk 2]
```diff
@@ -144,6 +145,41 @@ public void testallocateContainerDistributesAllocation() throws Exception {
...
+      }, executors.get(i)); futureList.add(future);
+    }
+    try {
+      CompletableFuture
+          .allOf(futureList.toArray(new CompletableFuture[futureList.size()]))
+          .get();
+    } catch (Exception e) {
+      Assert.fail("testAllocateBlockInParallel failed");
+    }
+  }
```
Comment (jiwq) [inline hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/TestSCMContainerManager.java]: Need a blank line between two methods.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 28 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 2 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 394 | trunk passed |
| +1 | compile | 203 | trunk passed |
| +1 | checkstyle | 50 | trunk passed |
| +1 | mvnsite | 0 | trunk passed |
| +1 | shadedclient | 815 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 125 | trunk passed |
| 0 | spotbugs | 238 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 416 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 394 | the patch passed |
| +1 | compile | 208 | the patch passed |
| +1 | javac | 208 | the patch passed |
| -0 | checkstyle | 31 | hadoop-hdds: The patch generated 2 new + 0 unchanged - 0 fixed = 2 total (was 0) |
| +1 | mvnsite | 0 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 662 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 127 | the patch passed |
| +1 | findbugs | 423 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 131 | hadoop-hdds in the patch failed. |
| -1 | unit | 955 | hadoop-ozone in the patch failed. |
| +1 | asflicense | 37 | The patch does not generate ASF License warnings. |
| | | 5170 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdds.scm.pipeline.TestRatisPipelineProvider |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-826/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/826 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 13d5d699ed50 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 4cb3da6 |
| Default Java | 1.8.0_212 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-826/2/artifact/out/diff-checkstyle-hadoop-hdds.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-826/2/artifact/out/patch-unit-hadoop-hdds.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-826/2/artifact/out/patch-unit-hadoop-ozone.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-826/2/testReport/ |
| Max. process+thread count | 4557 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/server-scm U: hadoop-hdds/server-scm |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-826/2/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerStateManager.java [hunk 3]
```diff
@@ -296,10 +296,10 @@ ContainerInfo allocateContainer(
         .setReplicationFactor(pipeline.getFactor())
         .setReplicationType(pipeline.getType())
         .build();
-    pipelineManager.addContainerToPipeline(pipeline.getId(),
```
Comment (jnp) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerStateManager.java]: I see that this method (allocateContainer) is also called from places where lock on the pipeline is not held.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/SCMContainerManager.java [hunk 4]
```diff
@@ -386,18 +386,17 @@ public ContainerInfo getMatchingContainer(final long sizeRequired,
...
-        synchronized (pipeline) {
+        containerIDs = getContainersForOwner(containerIDs, owner);
+        if (containerIDs.size() < numContainerPerOwnerInPipeline) {
           // TODO: #CLUTIL Maybe we can add selection logic inside synchronized
           // as well
-          containerIDs = getContainersForOwner(
-              pipelineManager.getContainersInPipeline(pipeline.getId()), owner);
           if (containerIDs.size() < numContainerPerOwnerInPipeline) {
             ContainerInfo containerInfo =
                 containerStateManager.allocateContainer(pipelineManager, owner,
```
Comment (jnp) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/SCMContainerManager.java]: There is one more place where containerStateManager#allocateContainer is being called from, where pipeline lock is not acquired.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerStateManager.java (hunk 3 above)
Comment (jnp) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerStateManager.java]: Should we also add a comment that this method must be called with lock held on the pipeline?

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/SCMContainerManager.java (hunk 4 above)
Comment (bshashikant) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/SCMContainerManager.java]: The SCM cli option will exercise this code path, which has an option to create the container. I think we can disable createContainer call from SCMCLI as it does not get used. What do you think?

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerStateManager.java (hunk 3 above)
Comment (bshashikant) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerStateManager.java]: Addressed in the latest patch.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/SCMContainerManager.java (hunk 4 above)
Comment (bshashikant) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/SCMContainerManager.java]: Ensured the pipeline lock is held on all places where allocateContainer is called.

**Pull request closed without merging**